### DISCUSSION TITLE
signinanonymously and captcha

### DISCUSSION THREAD

**Discussion by SgtErnestBilko** [supabase/supabase#29396] [Questions]
I have just added captcha (cloudflare - turnstile)  for user logins .    I also had a POST api route which received and used email data from SendGrid.   In order for this to work I used  signinanonymously.

As soon as I enabled captcha I got 'Error: Captcha verification failed'.

The Supabase documents say that 'It is recommended to set up captcha for anonymous sign-ins to prevent abuse. You can pass in the captcha token in the options param.'

OK but how do I generate a captcha token in an 'anonymous' POST request function without a cf-turnstile-response or CF-Connecting-IP ? 

Thank you

**Comment by J0** [comment]
Hey @SgtErnestBilko,

Thanks for trying our platform and flagging the issue. Anonymous sign ins should work with CAPTCHA client side. You can use `signInAnonymously({options:{ captchaToken}})` or similar. If you wish to use CAPTCHA with anonymous sign ins server side I think you'll need to forward the CAPTCHA token and response from the client ot the browser and then submit the token from the server side route.

**Reply by SgtErnestBilko** [reply]
Thank you for replying, I appreciate it.  "If you wish to use CAPTCHA with anonymous sign ins server side I think you'll need to forward the CAPTCHA token and response from the client ot the browser and then submit the token from the server side route."  Is the bit that is beyond my coding ability.  My server side component is a POST function within a route.ts.  It is pure server side in that it works independently of any user logins.   Parsed email data is POSTed to this route by SendGrid.  This data is used to update tables in a supabase db.  So, as far as I can see, there is no client side from which to get a token.

**Reply by silentworks** [reply]
@SgtErnestBilko you should show some code as that helps anyone trying to help understand your issue easier.

**Reply by SgtErnestBilko** [reply]
Fair comment (thank you)

```
"use server";

import { NextRequest, NextResponse } from "next/server";
import { createClient } from "@/utils/supabase/server";

export async function POST(request: NextRequest) {
  const supabase = createClient();

  try {
    // Validatation of data received goes here

    // Authenticate anonymously with Supabase
    const {
      data: { session },
      error: authError,
    } = await supabase.auth.signInAnonymously({
      // options: {
      //   captchaToken: turnstileResult.token,
      // },
    });
    if (authError) {
      console.error(
        "Error in (auth)/parse/route/POST/signInAnonymously: ",
        authError,
      );
      throw "error";
    }

    // Ensure we're using the authenticated client for subsequent operations
    const authenticatedSupabase = createClient();

    // uses authenticatedSupabase to call functions to operate on data

  } catch (error) {
    console.error("Caught error in (auth)/parse/route/POST: ", error);
    return NextResponse.json({
      status: 500,
      text: "Error in email parse",
    });
  }

  // Return a success response if everything went well
  return NextResponse.json({
    status: 200,
  });
}
```

**Comment by SgtErnestBilko** [comment]
I wonder is what I am trying to do actually possible ?

**Comment by SgtErnestBilko** [comment]
I think the answer is (but I could be wrong).......

Turnstile is for client components only.  In fact I think I read that using it in server components may breach turnstile terms of use.

I've decided to use 'signInAnonymously" without the captchas option as per my code above.  To make this work I have had to 'disable Bot and Abuse Protection - Enable Captcha protection'.  This doesn't appear to affect using turnstile for normal logins.  I have some protection on the POST route as it is rate limited.
